modules/artifact/src/main/java/org/mule/module/artifact/classloader/IBMMQResourceReleaser.java

```diff
@@ -0,0 +1,496 @@
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.module.artifact.classloader;
+
+import org.mule.runtime.module.artifact.api.classloader.ResourceReleaser;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.management.MBeanRegistrationException;
+import javax.management.MBeanServer;
+import javax.management.MalformedObjectNameException;
+import javax.management.ObjectInstance;
+import javax.management.ObjectName;
+import java.lang.ref.Reference;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Hashtable;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.stream.Collectors;
+
+import static java.beans.Introspector.flushCaches;
+import static java.lang.management.ManagementFactory.getPlatformMBeanServer;
+import static java.util.Objects.isNull;
+import static java.util.Objects.nonNull;
+
+
+public class IBMMQResourceReleaser implements ResourceReleaser {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(IBMMQResourceReleaser.class);
+  private static final String JMSCC_THREAD_POOL_MASTER_NAME = "JMSCCThreadPoolMaster";
+
+  private final static String THREADLOCALS_FIELD = "threadLocals";
+  private final static String INHERITABLE_THREADLOCALS_FIELD = "inheritableThreadLocals";
+  private final static String THREADLOCAL_MAP_TABLE_CLASS = "java.lang.ThreadLocal$ThreadLocalMap";
+  private final static String JUL_KNOWN_LEVEL_CLASS = "java.util.logging.Level$KnownLevel";
+  private final static String IBM_MQ_MBEAN_DOMAIN = "IBM MQ";
+  private final static String IBM_MQ_WORK_QUEUE_MANAGER_CLASS = "com.ibm.msg.client.commonservices.workqueue.WorkQueueManager";
+  private final static String IBM_MQ_ENVIRONMENT_CLASS = "com.ibm.mq.MQEnvironment";
+  private final static String IBM_MQ_COMMON_SERVICES_CLASS = "com.ibm.mq.internal.MQCommonServices";
+  private final ClassLoader driverClassLoader;
+
+
+  @Override
+  public void release() {
+    LOGGER.debug("Releasing IBM MQ resources");
+    //TODO add property to avoid releasing this resources
+    removeMBeans();
+    cleanJULKnownLevels();
+    cleanMQCommonServices();
+    cleanMQEnvironment();
+    cleanJmsTls();
+    cleanTraceController();
+    removeThreadLocals();
+  }
+
+  /**
+   * Creates a new Instance of IBMMQResourceReleaser
+   */
+  public IBMMQResourceReleaser(ClassLoader classLoader) {
+    this.driverClassLoader = classLoader;
+  }
+
+  /**
+   * Disposes all the IBM MQ related threads.
+   */
+  public void disposeMQThreads() {
+    for (Thread thread : getAllThreads()) {
+      if (thread.getName().equals(JMSCC_THREAD_POOL_MASTER_NAME)) {
```

Comment:
shoudl this be equals or a regexMatch? What happens if I deploy 10 apps, all using the connector? Won't I get 100 different threads with different (yet similar) names?

Reply:
This method is already present in the connector code. I removed it because it was not used by this class.